### ISSUE TITLE
RFC: Change to Directory Structure

### ISSUE THREAD

**Issue by joshdmiller** [ngbp/ngbp#356]
Since authoring `ngbp` several years ago, the way I structure directories has evolved. The philosophy is still the same, principally:
- Organize by feature, not by layer
- Maintain a distinction between app-specific code and reusable code

However, there are a few changes that I've found helpful. I'd like to discuss changing the directory structure to something like this:

```
ngbp/
    |-- src/
        |-- index.html
        |-- ngbp.js
        |-- ngbp.spec.js
        |-- states/
            |-- states.js
            |-- states.spec.js
            |-- home/
                |-- home.js
                |-- home.spec.js
                |-- home.tpl.html
                |-- home.scss
                |-- home.scenario
            |-- about/
        |-- component1/
            |-- component1.js
            |-- component1.spec.js
            |-- component1.tpl.html
            |-- component1.scss
        |-- component2/
    |-- static/
    |-- vendor/
```

There are two key distinctions:
1. Style is inline with the component it's styling
2. `app` and `components` are deprecated in favor placing app-specific code in `src/states/` to make it clear that it is _state-specific_ (e.g. _route-specific_) code that goes there (formerly `src/app`) and that all other code is just in `src/`. In essence, `states` becomes a component in and of itself.

The module structure would similarly mirror the filesystem:

```
ngbp
  ngbp.states
    ngbp.states.home
    ngbp.states.about
  ngbp.component1
  ngbp.component2
```

I've used this to good effect in several projects. What are the thoughts of the community?

[Send us your reckons](https://www.youtube.com/watch?v=OQnd5ilKx2Y).

**Comment by Merott** [comment]
It looks good to me, and in fact that's how I've also been structuring my projects recently. I just call the folder `areas` in place of `states`. I also tend to put the directives/services/resources into their own top-level folders under `src`, but your suggested structure fully supports that anyway.

So the only thing I'd say is to make sure that the folder names are as configurable as possible, which I'm sure is always part of the plan.

**Comment by mustela** [comment]
Hi there, in MHO I would suggest making a main `components` folder and add in there all the components (directives, services, etc). Since in a large app you could easily have hundred of them and `states` could be easily "lost". It would help to really separate things, you could even have a "states" component in your app with no relationship with the `ngbp.states`

```
states
  home
  about
components 
  component1
  component2
```

**Comment by Merott** [comment]
@mustela I think that's not going to be a problem as you can structure content under `src` however you like. I also agree with your approach, and that otherwise `states` would easily get lost in a myriad number of other components.

**Comment by joshdmiller** [comment]
I agree with @Merott that such a structure would inherently support virtually any configuration, from ones that are debatably very similar (such as those we're discussing here) to ones that I would not recommend, such as organizing by layer.

The important thing to note is that there are fundamentally two complementary issues to ngbp: (a) the build tool and (b) the recommended structure. Right now, the build tool is somewhat confined by its recommended structure; under this proposal (or a similar one), we would remove that limitation. Apart from the build tool, we would have to decide what _recommendation_ we want to make about file structure.

For that question, my feedback on @mustela's suggestion is that I understand and appreciate the concern with `states` getting lost, but I think nesting components in a subdirectory would lead either (a) to an awkward module structure (e.g. `ngbp.components.component1`), or (b) to the module structure not matching the file system. I would prefer the former to the latter, if we were to nest the components.

What are your thoughts?

**Comment by dancancro** [comment]
For what it's worth, [here](http://www.dancancro.com/app-file-structure-comparison/) is a report I made to classify the code in an application and to show side-by-side where it lives in applications produced by some respected experts.  Or if that's confusing, row 127 of [this big spreadsheet](https://docs.google.com/spreadsheets/d/1r8rJy2Q5p5QORYKcye93UECwOlSgFL24c5fyF7dqhaM/edit#gid=1607194899) contains a slightly different take on it.

I recently added Ward Bell's Pizza app demo.  If anyone wants to add a column for ngbp present or future, let me know and I'll make you an editor.

Not saying that these ways are better, but IMHO, this kind of decision should not be made without full knowledge of all the ways that other smart people do things.  Also, the report contains around 50 different types of code, each with a few classifications for a total of around 130 classifications.  

A good solution should have well-considered, unambiguous answers to all ~130 questions of "where should I put code of this classification?".

**Comment by joshdmiller** [comment]
Hey Dan. There's actually already an ngbp column in that spreadsheet (though it's not very accurate), but as I've expressed in emails before, I don't see a lot of value in that level of comparison. Fundamentally, I feel it's comparing incomparable products. mean.io and ngbp have completely different goals and a completely different audience.

But back on the subject at hand, I see great utility in reaching community consensus on best practices for AngularJS. When we pare down your spreadsheet to the AngularJS goodness (eliminating the backend, other frameworks, etc), we see there is in fact broad consensus on the big picture. For example, while there are a few exceptions, most thought leaders want JavaScript files organized by feature and most want diverse file types side-by-side with the component that defines them. I don't think the community disagrees on any of the important things (by and large). As another example, I personally prefer `FooController`, but have absolutely no objection to someone using `FooCtrl` or `FooCntrlr`.

In the same proverbial vein, I also do not think it worthwhile to come up with recommendations at the detail of "130 classifications" of "50 different types of code". If we agree on the broad strokes and the build tool is flexible enough, developers should have the freedom to decide the minutiae for themselves, if for no other reason than as a kickstarter ngbp is never going to have 50 different types of code! :-)

However, I opened this issue with the express intent of soliciting feedback on this structure, and I appreciate you chiming in. If my proposal doesn't match what you (or anyone else) thinks a good AngularJS project should look like, I wholeheartedly, emphatically, indubitably, irrefragably, undeniably, and indisputably would love to hear it. Nothing would make me happier than to elicit some lively debate on this topic.

Do you see a way to improve what I've outlined here?

**Comment by mustela** [comment]
@Merott @joshdmiller I know the structure is really flexible, and actually you can create what ever you want under `/src` but ngbp should _suggest_ a way of doing it, not telling people where to put your non-states code, would rise a lot of questions I think.

Also @joshdmiller I'm not sure about the problem with having `ngbp.components.component1` would you mind to clarify, why do you think its awkward? Also about _(b) to the module structure not matching the file system_, why do you think that?

I keep thinking that having `/states` in the same level of everything else, won't be so clear.

**Comment by joshdmiller** [comment]
@mustela I strongly agree that ngbp should recommend an approach (and implement it to the degree necessary in the sample code).

I mean the path to the module definition should match its file system path, as much as possible. Since all code is in `src/`, a module like `myApp.richtext.toolbar` should be found in `src/richtext/toolbar`. If we place reusable components in a subdirectory, then it would have to be `myApp.components.richtext.toolbar` to be consistent. Similarly, I see `myApp.states` as the top-level module under which a route/state/area sits. For example, `myApp.states.contacts.detail`.

As far as what the _problem_ is with having components within the extra `components` submodule, there isn't one. I am just not sure it is necessary. However, I definitely see your point and (due to the flexibility of the build tool) I'm happy to defer to the community here if I'm outnumbered (as I seem to be so far).

**Comment by mustela** [comment]
@joshdmiller thanks for giving more details, got your point now about the file system path. But don't see a real issue in adding the `components` and `states` modules to match them. I believe it keeps things organised and clear. 

But doing that change would end up with the [same current structure](https://github.com/ngbp/ngbp/tree/v0.3.2-release/src), which I believe is a nice one. What I would change is the `common` dir to be the `components` one. Since I don't really see how you share those _common_ components through applications. If I need a _drag and drop_ module, I code it in a separated repo and include it using `bower` for instance, not copying the files on each project. Maybe I'm missing something here? 

So maybe you don't really need to change the entire dir structure after all...

**Comment by joshdmiller** [comment]
As an humorous aside, the `common` directory used to be called `components` until the community requested it be changed. :-)

But the change I am proposing is still necessary, even if we keep the `components/` subdirectory. Right now, there are separate globs for components and app (in some cases), and I am recommending that those be eliminated in favor of a single set of globs for everything in `src/`, while recommending the `states` and components approach by example. In addition, styles should be included alongside component code, eliminating the `src/less` directory entirely, and moving `src/assets` to `static/` in the root of the project. Essentially:
- JS: `src/**/*.js`, `!src/**/*.spec.js`
- Specs: `src/**/*.spec.js`
- Templates: `src/**/*.tpl.html`
- HTML: `src/**/*.html`, `!src/**/*.tpl.html`
- LESS: `src/**/*.less`
- Assets: `static/**/*`

**Comment by mustela** [comment]
Thats a good one @joshdmiller! is there any thread about the _components vs common_ discussion ? 

About the change you are requesting of having a single set of globs, I don't see an issue with that, so +1

**Comment by joelhooks** [comment]
How do you handle the less/sass imports with them alongside?

**Comment by dancancro** [comment]
My remarks, written with love...

You wrote:

> Hey Dan. There's actually already an ngbp column in that spreadsheet (though it's not very accurate), but as I've expressed in emails before, I don't see a lot of value in that level of comparison. Fundamentally, I feel it's comparing incomparable products. mean.io and ngbp have completely different goals and a completely different audience.

I meant adding a column for ngbp to the [file structuring comparison](https://docs.google.com/spreadsheets/d/1fOvWxAKEN_ilcGdkxW5pkW1RSsYckxRihtBQv2wW69Y/edit#gid=0), not the [starters comparison](https://docs.google.com/spreadsheets/d/1r8rJy2Q5p5QORYKcye93UECwOlSgFL24c5fyF7dqhaM/edit#gid=1534967054).

Label them as you will, but ngbp and mean.io are the same in one important way.  When someone is about to start a project not from scratch they must choose either ngbp or mean.io or some other at the expense of not choosing every other alternative.  They are substitutes and this is the point of the comparison - helping people decide which ONE to use because they can only choose ONE.

You wrote:

> (though it's not very accurate)

You are more than welcome to correct any inaccuracies you find.  It wouldn't take much time.  You have editor permission already.  I don't really understand why you'd see an inaccuracy and just leave it there.  This phenomenon is really puzzling to me.  A few people (Dustin, Balthazar, Mark) have been happy to make their columns accurate and their projects comparable to ease the distress of the poor, tortured masses faced with such a hard choice, but most have not.  In all honesty, trying to understand why the others have chosen not to update their project's facts is making me a bit crazy.  Is there anything I can do to make doing so more worth your while?  You will be helping people to save so much time and anguish, and you won't even need to make a sales page for your own project.  Just point them to the spreadsheet and they can see all of its benefits :)

You wrote:

> developers should have the freedom to decide the minutiae for themselves, if for no other reason than as a kickstarter ngbp is never going to have 50 different types of code! :-)

WIth all due respect, I just don't understand this.  Freedom of choice is only a value in things that apply to the five senses: taste, touch, smell, sound and sight, and belief systems that cannot be proven - but not to things that are functional.  The value of such things can be measured in units of time saved which can be saved in any number of ways.  For example, "preference for coffeescript" is a linguistic error.  If giving up coffeescript means that you suddenly have access to a thousand typescript modules which in turn saves you a hundred hours, then giving up coffeescript is what you need to do.  You did not prefer coffeescript, you just calculated that its use saved you more time than its non use.  Same with putting a file in one place or another.  Freedom of choice is not applicable here because it does not apply to any of the five senses.  I don't understand why developers should have this thing instead of the time savings they really want and could get through standardization which would let them simply pop in compliant code and have it work.  It's like granting people in Manhattan their choice of farm tractor.

Why can't an app started with ngbp eventually have every type of code?  [Have another look at them](https://docs.google.com/spreadsheets/d/1fOvWxAKEN_ilcGdkxW5pkW1RSsYckxRihtBQv2wW69Y/edit#gid=0).  None of them are at all obscure -  are unit tests obscure?  are images obscure?  If such a limitation exists in ngbp, I'm pretty surprised by this. 

Please come to the next hangout!  We miss you and will be talking about directory structures.

**Comment by joshdmiller** [comment]
I appreciate the feedback. Here are some more thoughts. :-)

> I meant adding a column for ngbp to the file structuring comparison, not the starters comparison.

Sorry, I misinterpreted your original post. However, if you grant me write access to the file structure sheet, I'll add ngbp. :-)

Similarly, I'll attend the next hangout; I appreciate the renewed invitation.

> This phenomenon is really puzzling to me. A few people ...  have been happy to make their columns accurate ... but most have not.

As I said, it's honestly because I have limited time and I don't see value in the comparison.

As a consultant, if someone walked up to me and said, "we're starting a project in AngularJS and we're probably running Node on the backend and MongoDB to house our data. Should we use ngbp?" My response would be this: "ngbp would certainly work for you, but if you have a full-stack team and plan on running a single repository, then mean.io will probably be a little more helpful in the beginning."

My point is that the many assumptions in the above conversation had nothing to do with 95% of the comparisons the spreadsheet makes. Before we got into any talk about file structures or naming conventions, they've outlined a full stack, front end to server to database; they've outlined their general workflow (one repo for front and back); and they've identified their team as full-stack. If _any_ of those assumptions don't hold, mean.io might not be the best choice - **all other things aside**. I personally feel the spreadsheet is too low-level to make valuable comparisons from projects with divergent goals.

Further, I think the comparison is unfair. It makes value judgments _a priori_ about what is a "benefit", then assigns weighted "points" to each benefit, and then gives a "score" to a project based on its accumulation of said benefits, even going so far as color-coding in red where a project falls short. Here is a brief list of "benefits" with which I disagree:
- being a generator (though I understand some find this valuable) gets 50x more points than a starter
- using any framework that's not angular (e.g. ember, backbone, etc.)
- using a library (e.g. bootstrap, ionic, etc.)
- including an API server
- setting up a database
- having a Facebook page
- splitting components by layer gets 50% more points than just organizing components well
- prefixing database table names
- not pluralizing words (!?)
- many more...

These "benefits" mix starters with generators with style guides with best practices. How would you rate ngbp on "no business logic in controllers" or its "approach to AngularJS directives"? I agree with the former and certainly have opinions on the latter, but since ngbp is a kickstarter, it is a _non sequitur_ to say it doesn't have those "benefits".

To summarize: if someone who is new ng development looks at this spreadsheet to decide which kickstarter/generator/reference app/style guide to use, I don't think she'll walk away with a true understanding of their differences (see above about ngbp vs mean.io for an example).

All that said, I appreciate all the hard work you've put into this and am sorry you feel frustrated by my lack of participation in that spreadsheet. So, I will update the ngbp column. :-)

> Freedom of choice is not applicable here... . I don't understand why developers should have this thing instead of the time savings they really want and could get through standardization ... .

This is reductionistic and absolutist. I've used angular with JS, CoffeeScript, and ES6; in each case, the team chose what was right for the project. My colleagues have used TypeScript or DART in their angular projects to great success.

I personally happen to think using ES6 for some syntactical sugar (particularly fat arrows, `let`, and classes) leads to more legible code and is more future-proof than CS or TS, but I run _teams_. If my team doesn't know ES6, I am forced to weigh the cost (both financial and cultural) of demanding that transition against its benefits. Sometimes I've made that choice; other times I have not. It was never proverbially "black and white".

This also ignores the value of personal preference. Some languages make more sense to some people than to others. These developers will be more productive in the language they grok than the language they tolerate (definitively in the short term, arguably in the long).

As thought leaders, we should definitely say on what we agree and with what fervency we hold those positions. I feel much stronger about organizing by feature than I do about naming conventions. However, for a trivial application like a To Do list, organizing by feature would provide arguably very little benefit while still carrying the same cognitive load. In such a case, I would not fault a developer for choosing to organize by layer - or even to organize all in one file!

Bottom line: not all projects are the same - and even when they are, it's okay to disagree on some things.

> Why can't an app started with ngbp eventually have every type of code?

My point wasn't that there _couldn't_ be many types of files only that the differences between them were largely irrelevant. ngbp says "organize all feature code into discrete modules". What is "feature code"? Scripts, styles, templates, and tests. It also says everything else is "static" and we don't care about it, so "throw it in this folder and we won't touch it". That's it. I don't care whether that feature is a service or directive or filter or state or whatever, because it is my opinion they should be treated the same: as discrete modules. Similarly the distinction between images and fonts: they're both static assets, so we treat them the same way.

Most other things depend on the nature of the product. If one needs a favicon, she can add it as a static asset and update `index.html` accordingly. Similarly, ngbp runs a very basic set of jshint rules; if a team wants to alter them, they can alter them! If they have a lot of deviations from the norm, they can create a `jshint` file per the build tool's instructions. This is web development 101.

We can't possibly need to spell out how to handle 100+ different types of files. Honestly, what am I missing here?

---

I would sum up my opinion like this: We should build our tools for _engineers_, not for _code monkeys_. So let's assume they have some intelligence and problem solving skills and access to Google, then draw the broad strokes, give some collective wisdom, provide a tool that will work with whatever choices they ultimately make, and foster a community for sharing expertise. But let's not treat developers like children.

**That's the goal of ngbp.**

**Comment by dancancro** [comment]
Great.  I'm glad you can make it to the hangout.  

Some more comments...

> if you grant me write access to the file structure sheet, I'll add ngbp. :-)

I moved the file structures spreadsheet into a new page of the starters one.  So the permissions will apply the same way.  Thanks for your offer to describe ngbp's structure.  Hopefully it won't take too much time.  Just do a `find . *.*` and paste in things where they fit best or stick them in 'other'.

> As a consultant, if someone walked up to me and said, "we're starting a project in AngularJS and we're probably running Node on the backend and MongoDB to house our data. Should we use ngbp?"

If you hover over the cells of the second column you will see lists of starters that offer the given benefit.  This would be useful to people with dealbreakers - benefits that absolutely must be provided.  My agenda here on earth is to reduce the choices and get people to work together on something amazing, not to make it easier for people to pick among a huge array of half good choices.  So this is all I want to offer the dealbreaker people for now.  A business stakeholder will give up MongoDB if doing so will save money.  Same with any other sacred cow.

> I personally feel the spreadsheet is too low-level to make valuable comparisons from projects with divergent goals.

I've been studying this stuff for a while and I still don't clearly understand what the difference in goals is.  Anyone who uses any of these things wants to make a web application with some set of functionality to it.  And they sometimes have platform or language loyalties.  The starters are all general purpose and to my knowledge, that's as specific as their audience definitions get.

>  I think the comparison is unfair

The point values are totally dependent on my beginner appraisals which are far from perfect.  I give more thought to the summary totals than the individual point values.  I'd **love** to confer about this with experts and improve them, but here's the main thing... **as imperfect as it is, there is no more omniscient, unbiased alternative way to choose which thing to use**.  The status quo decision process is to rely on the accident of whoever you happen to ask for advice or whichever blog you happen to read.  I add things to the list anytime I read a reason why someone likes a particular option.  I'm very inclusive about this.  So if there's anything about ngbp that you like but isn't listed, just let me know.

You can also get around the subjective part by going to the Differences page and pointing to a cell that compares two things.  A popup will appear that just lists the trade-offs which are objective.  Ignore the numbers, but the lists of tradeoffs are factual - A has these things B doesn't have and B has these things A doesn't have.  I could have included this without having any points or scores or colors, but even doing this much takes more time than people want to spend and tilts a typical human brain.

I want to make a little app to make these tradeoff lists more readable, but this is there for now.  I also want to make an app that lets users input their own point values to bear on the existing set of benefit facts.  That would be great but only really useful for experts because beginners are clueless about how relatively important things are.  

> But let's not treat developers like children.

People are free to ignore your answers to the question "where should I put this kind of code?", but you can still give them.  I don't see any insult in doing that.  I am an adult and consider it a service for someone to relieve me of the burden of choosing between two equal options so I can think about other things that do matter.

Sorry this was so long.  See you soon!

**Comment by aa-telali** [comment]
This might be a noob question, I apologise in advance..

The main issue is that states/ isn't a component and could either clash with a 'states' component or become difficult to find in a long list of folders - Would it not be better to prefix it with an underscore or something? That way it remains distinguishable, appears at the top of the list and is less likely to clash with a component name.

**Comment by aa-telali** [comment]
Just had a look at the angular/angular-seed project and an ionic generator..

They tend to have a components/ sub-folder (as @mustela suggested) but no states/ folder, so all view folders are in the root (src/home src/about etc.). That actually makes more sense to me and it looks more like a standard website.

Other observations:
- Most people tend to use app/ instead of src/ although I understand the distinction you're trying to create (source as opposed to output).
- Continue to use the standard term assets/ instead of static/ although, again static/ is more correct, but assets/ is used by most generators and is familiar and therefore, easier to migrate existing projects to.

In this case the folder structure would be like this:

```
[ngbp]/
    |-- app/
        |-- index.html
        |-- [ngbp].js
        |-- [ngbp].spec.js
        |-- home/
            |-- home.js
            |-- home.spec.js
            |-- home.tpl.html
            |-- home.scss
            |-- home.scenario
        |-- about/
        |-- components/
            |-- components.js
            |-- components.spec.js
            |-- component1/
                |-- component1.js
                |-- component1.spec.js
                |-- component1.tpl.html
                |-- component1.scss
            |-- component2/
    |-- assets/
    |-- vendor/
```

Am I right in thinking [ngbp] will be replaced by the app-name?

In conclusion:
- If you're trying to lead on the 'best' way to name folders then keep src/ and static/.
- If you pushing a 'better' way then stick to app/ and assets/.
- I push for no states/ folder (which is confusing) and a components/ subtree instead, which could still clash with a components/ state but is likely to be renamed in that case as it's there purely for organisation as opposed to being part of the build process right?

Oh, and thank you Josh for creating ngbp. It's taught me a lot. :-)
